File: scala/src/main/scala/leetcode/p0001/TwoSum.scala
package leetcode.p0001

import scala.collection.mutable

object TwoSum:
  /**
   * Returns the indices of two numbers whose sum equals the target.
   *
   * @param nums   input numbers
   * @param target required sum
   * @return the two matching indices
   */
  def solution(nums: Array[Int], target: Int): Array[Int] =
    val complementIndexMap = mutable.Map.empty[Int, Int]
    for ((value, index) <- nums.zipWithIndex) {
        val complementValue = target - value
        if (complementIndexMap.contains(complementValue)) {
            return Array(complementIndexMap(complementValue), index)
        } else {
            complementIndexMap(value) = index
        }
    }

    return Array.empty[Int]

File: java/lib/src/main/java/leetcode/p0001/TwoSum.java
package leetcode.p0001;
import java.util.HashMap;

public class TwoSum {
    public int[] solution(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int index = 0; index < nums.length; index++) {
            int complement = target - nums[index];
            if (map.containsKey(complement)) {
                return new int[] { map.get(complement), index };
            } else {
                map.put(nums[index], index);
            }
        }

        return new int[] {};
    }
}
